### PR TITLE
CodeMirror 3.0 support

### PR DESCRIPTION
**Pull request by douglasduteil**
### Update the CodeMirror version supported from 2.3 to 3.0. #312

Add access to the event 
- cursorActivity 
- viewportChange 
- gutterClick 
- focus 
- blur 
- scroll 
- update

With the same way of  `onChange` : 

``` html
 <textarea 
ng-model="input"
ui-codemirror="{
   onChange:         doChange ,
   onCursorActivity: doSomething
}" 
>
```

Demo : http://jsbin.com/acitel/1/

### CODE REVIEW

File: modules/directives/codemirror/codemirror.js [hunk 1]
```diff
@@ -2,77 +2,70 @@
...
-        if (codemirror) {
-          codemirror.toTextArea();
-        }
-        // Create the new codemirror widget
-        codemirror = CodeMirror.fromTextArea(elm[0], options);
-      };
 
-      // Initialize the code mirror widget
-      updateCodeMirror(uiCodemirrorGet());
+			events = "cursorActivity viewportChange gutterClick focus blur scroll update".split(' ');
```
Comment (ProLoser) [inline modules/directives/codemirror/codemirror.js]: Why not just make this an array to begin with? Also, might as well declare it up around line 7

Comment (douglasduteil) [inline modules/directives/codemirror/codemirror.js]: Yep. The same with `options` [line 18](https://github.com/angular-ui/angular-ui/pull/314/files#L0R18) ?

Comment (ProLoser) [inline modules/directives/codemirror/codemirror.js]: No, attrs isn't available earlier up in the execution chain, but the events array never changes and doesn't need to be created on every linking pass.

Comment (douglasduteil) [inline modules/directives/codemirror/codemirror.js]: Yes `opts` changes for each element 
line 19 :  

```
opts = angular.extend({}, options, scope.$eval(attrs.uiCodemirror));
```

But what about `options`  that depends of the `uiConfig`
 line 18:

```
options = uiConfig.codemirror || {};
```

Is it use whatever the element during compile-time ?

**Comment by ProLoser** [comment]
Needed to review this more before I merged it.

Why are you deferring the initialization? You did not explicitly state in comments the reason for this, and it's not good to simply use this as an end-all solution needlessly.

**Comment by douglasduteil** [comment]
Yep sorry I'll add some comments ...
Basically, I do it to let the compiler go through the whole node. When added CodeMirror takes textarea.nextSibling that isn't compile yet. 
From [codemirror.js#L3179](https://github.com/marijnh/CodeMirror/blob/master/lib/codemirror.js#L3179) : 

``` javascript
textarea.parentNode.insertBefore(node, textarea.nextSibling);
```

That why it's doing strange things (see http://jsbin.com/aqecex/6)

**Pull request merged**